### PR TITLE
feat(votes): add allow abstain option and abstention results (GH-1470)

### LINKED ISSUE #1470: Allow organizers to enable abstaining when creating a vote
## Problem

Vote organizers cannot offer voters the option to abstain. When creating a vote in LFX One there is no setting to enable abstention, so every ballot forces a choice between the listed options even when the governance workflow expects members to be able to formally abstain. The voter-facing ballot already knows how to render an "Abstain" option and the results view already displays abstentions — but those only appear when a vote was created with abstention enabled, which is impossible through the UI today.

## Outcome

Organizers can turn on an "Allow abstain" option when creating or editing a vote. When enabled, voters see an Abstain choice on their ballot (both the vote drawer and the dashboard inline ballot) and results correctly reflect abstentions. When disabled, ballots behave exactly as they do today. Editing a vote shows and preserves the current setting.

## Success criteria

- The vote creation form includes a visible control to enable or disable abstention
- The setting can be reviewed before the vote is opened
- Votes created with abstention enabled show the Abstain option to eligible voters
- Votes created without it behave exactly as before
- Editing a vote reflects the current abstain setting and does not silently reset it

### PR DESCRIPTION
**Pull request by audigregorie**
## Summary

Adds an "Allow Abstain" setting to vote creation so organizers can let participants abstain instead of choosing an option, and surfaces abstention counts in the results drawer. The flag is threaded end-to-end — form, request builders, review step — so editing a vote never silently resets it, and results for abstain-enabled votes now show how many voters abstained alongside the usual participation stats.

Resolves #1470

### Vote creation

| Before | After |
|---|---|
| There was no way to let voters abstain — every participant had to pick an option, and nothing about abstaining appeared anywhere in the create/edit flow. | A Yes/No "Allow Abstain" select (defaulting to No) appears in the basics step of the vote form, and the chosen value is shown as a read-only row on the review step before submitting. |

### Vote results

| Before | After |
|---|---|
| The results drawer showed only eligible voters, responses, and response rate — abstentions were invisible even though the backend counts them. | For votes created with abstaining allowed, the participation card gains an "Abstained" count, and a muted italic "Abstain" row appears under the option list showing abstention count and share of all responses cast. Loading skeletons reserve space for both so nothing shifts on load. |

## Technical changes

`packages/shared/src/constants/poll.constants.ts` — Shared constants

- Adds `VOTE_ALLOW_ABSTAIN_OPTIONS` Yes/No options for the basics-step select, mapping to the `allow_abstain` field on the Vote interface

`packages/shared/src/interfaces/poll.interface.ts` — Shared types

- Adds `allow_abstain: boolean` to `VoteFormValue`
- Adds `abstainedVoters` and `abstainedRate` to `VoteParticipationStats`, documented as a subset of `totalResponses` (upstream counts abstentions as cast responses but excludes them from choice tallies)

`packages/shared/src/utils/vote.utils.ts` — Shared vote helpers

- Threads `allow_abstain` through all four vote request builders (`buildCreateVoteRequest`, `buildDraftVoteRequest`, `buildUpdateVoteRequest`, `buildDraftUpdateVoteRequest`) so ITX full-replace updates never silently reset the flag
- Maps `allow_abstain` back into the form in `mapVoteToFormValue`, defaulting to `false` for votes that predate the field
- Extracts the drawer's participation-stats computation into a shared `computeVoteParticipationStats` util, extended with `abstainedVoters` / `abstainedRate`

`packages/shared/src/utils/vote.utils.spec.ts` — Tests

- Covers the builders and mapper round-trip for `allow_abstain`, plus `computeVoteParticipationStats` edge cases (null results, zero eligible voters, abstention rates)

`apps/lfx-one/src/app/modules/votes/vote-manage/vote-manage.component.ts` — Vote form state

- Adds a non-nullable `allow_abstain` `FormControl` defaulting to `false`, and patches it when editing an existing vote

`apps/lfx-one/src/app/modules/votes/components/vote-basics/vote-basics.component.html, vote-basics.component.ts` — Basics step

- Renders the "Allow Abstain" `lfx-select` bound to the `allow_abstain` control with helper text and a `data-testid`

`apps/lfx-one/src/app/modules/votes/components/vote-review/vote-review.component.html, vote-review.component.ts` — Review step

- Adds a read-only "Allow Abstain: Yes/No" row backed by an `initAllowAbstain()` computed signal

`apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html, vote-results-drawer.component.ts` — Results drawer

- Replaces the inline participation-stats computed with the shared `computeVoteParticipationStats` util
- Renders the "Abstained" stat in the participation card and the muted "Abstain" row in both live and final results, gated on `voteData.allow_abstain`
- Adds matching skeleton placeholders for the new stat and row to avoid layout shift while loading

### CODE REVIEW

File: apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html [hunk 1]
```diff
@@ -681,6 +730,25 @@ <h4 class="text-sm font-medium text-slate-700">
                 </div>
               </div>
             }
+
+            @if (voteData.allow_abstain) {
```
[the remark this answers is not in the document]
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html]: Fixed in 28e578d0d — the abstention result row now renders in both voter final views (generic and ranked), and the zero-results state counts abstentions as cast responses instead of showing "No votes were submitted".

File: apps/lfx-one/src/app/modules/votes/components/vote-basics/vote-basics.component.html [hunk 2]
```diff
@@ -132,6 +132,23 @@ <h3 class="text-base font-medium text-gray-900 mb-2">{{ committeeLabel.singular
...
+              styleClass="w-full"
+              appendTo="body"
+              [form]="form()"
+              control="allow_abstain"
+              [options]="allowAbstainOptions"
+              optionLabel="label"
+              optionValue="value"
+              placeholder="Select allow abstain"
+              data-testid="vote-allow-abstain-select">
+            </lfx-select>
```
[the remark this answers is not in the document]
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-basics/vote-basics.component.html:150]: Fixed in 28e578d0d — the Allow Abstain select now passes inputId="allow-abstain" through the lfx-select wrapper, so the label is programmatically associated and focuses the control.

File: apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html [hunk 3]
```diff
@@ -393,6 +397,13 @@ <h2 class="text-base font-semibold text-slate-900">Vote Details</h2>
                 <p-skeleton width="100%" height="8px"></p-skeleton>
               </div>
             }
+            @if (voteData.allow_abstain) {
```
[the remark this answers is not in the document]
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html]: Fixed in 28e578d0d — the abstain skeleton row is now gated on isGenericPoll() as well, so ranked votes no longer reserve space for a row that never renders.

File: apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html [hunk 4]
```diff
@@ -306,6 +306,23 @@ <h4 class="text-sm font-medium text-slate-700">
                   </div>
                 </div>
               }
+              @if (voteData.allow_abstain) {
```
[the remark this answers is not in the document]
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html]: Fixed in 4dce2cf — both voter closed-view abstain rows (generic and ranked) are now gated on !voteResultsError(), so a failed results request no longer renders fabricated "0 abstentions" above the error message.

File: packages/shared/src/utils/vote.utils.ts [hunk 5]
```diff
@@ -120,11 +122,34 @@ export function mapVoteToFormValue(vote: Vote): VoteFormValue {
...
+export function computeVoteParticipationStats(results: VoteResultsResponse | null): VoteParticipationStats {
+  if (!results) {
+    return { eligibleVoters: 0, totalResponses: 0, participationRate: 0, abstainedVoters: 0, abstainedRate: 0 };
+  }
+
+  const eligibleVoters = results.num_recipients || 0;
+  const totalResponses = results.num_votes_cast || 0;
+  const abstainedVoters = results.num_abstained || 0;
+  const participationRate = eligibleVoters > 0 ? Math.round((totalResponses / eligibleVoters) * 100) : 0;
+  const abstainedRate = totalResponses > 0 ? Math.round((abstainedVoters / totalResponses) * 100) : 0;
```
Comment (ahmedomosanya) [inline packages/shared/src/utils/vote.utils.ts:148]: **Do `abstainedRate` and `option.percentage` share a denominator? (blocker if not)**

This divides by `num_votes_cast`, which includes abstentions. The choice-percentage fallback in `vote-results-drawer.component.ts` divides by the sum of `generic_choice_votes`, which excludes them:

```ts
// computePercentage
return Math.round((voteCount / totalVotes) * 100);
```

With 4 eligible voters, 3 responses, 1 abstention and 2 votes for A, the creator live view renders "A — 2 votes · 100%" directly above "Abstain — 1 abstention · 33%", and the two bars in one stack sum to 133%.

This only bites on the fallback path — if upstream's `generic_choice_votes[].percentage` is already computed over all responses cast, the numbers agree and there's nothing to fix. Could we confirm which base upstream uses? If it excludes abstentions, I'd suggest putting the abstain share on the same base:

```ts
const abstainedRate = abstainedVoters + choiceVotesTotal > 0
  ? Math.round((abstainedVoters / (abstainedVoters + choiceVotesTotal)) * 100)
  : 0;
```

Failing that, labelling it "33% of responses" would at least make the different base visible rather than implied.

File: apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html [hunk 6]
```diff
@@ -534,6 +586,23 @@ <h3 class="text-base font-semibold text-slate-900">Q{{ idx + 1 }}. {{ question.q
                 </div>
               </div>
             }
+
+            @if (voteData.allow_abstain) {
```
Comment (ahmedomosanya) [inline apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html]: **The creator-side abstain UI still shows zeros when the results request fails**

`voteResultsError` is set in creator scope too (`canFetch` is `audience === 'creator' || vote?.status === ENDED`), and `catchError` returns `of(null)`, so `computeVoteParticipationStats(null)` supplies zeros. The creator branches have no error branch, so this row paints "Abstain — 0 abstentions · 0%" as if measured. Same for the creator final row at line 767 and the "Abstained" stat at lines 474, 623 and 678.

This is the finding already fixed for the voter rows in 4dce2cf — could we apply the same `&& !voteResultsError()` gate to the creator equivalents?

```html
@if (voteData.allow_abstain && !voteResultsError()) {
```

I realise the stat card already zeroed participation figures there before this PR, so the creator error state is pre-existing; the new row is just the most assertive element of the set.

**Review by ahmedomosanya — COMMENTED** [review]
**LGTM overall, with two things I'd like to settle first** — left inline on the specific lines.

Threading `allow_abstain` through all four builders with a test asserting `'allow_abstain' in request` is exactly the right guard against the ITX full-replace reset, and moving the participation-stats computation into `computeVoteParticipationStats` with real edge-case tests is a clear improvement over the inline computed. It also closes a real loop: `vote-cast-drawer` and `vote-ballot-inline` already read `vote.allow_abstain`, so nothing was setting the flag they depend on.

**Non-blocking:**

- The ranked creator branch shows the "Abstained" stat but no abstain row, while the ranked voter closed view renders one. Intentional?
- The abstain block is duplicated four times (html lines 309, 365, 590, 767) with the pluralization repeated in each. This file already has the `#commentResultsList` + `ngTemplateOutlet` pattern for exactly this — lifting the row into an `ng-template` would mean the error gate below only has to be fixed once, and would have prevented the creator/voter divergence.
- `hasZeroVotes()` and the per-branch error gating are the likeliest bits to regress and live in a component with no spec. A drawer spec may be more than this PR should carry, but the `abstainedVoters > 0` short-circuit is cheap to cover.
- `placeholder="Select allow abstain"` is unreachable — the control is non-nullable with a `false` default, so "No" is always selected.
- For a multi-question vote, `num_abstained` is vote-level while tallies are per-question. Placing the row after all questions is right — but can a voter abstain on one question and answer another, or is abstention always whole-ballot? If per-question is possible, one trailing row will understate it.

File: apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.ts [hunk 7]
```diff
@@ -476,6 +471,8 @@ export class VoteResultsDrawerComponent {
 
   private initHasZeroVotes(): Signal<boolean> {
     return computed(() => {
+      // Abstentions are cast responses upstream but excluded from choice tallies — they must not read as "no votes".
+      if (this.participationStats().abstainedVoters > 0) return false;
```
Comment (ahmedomosanya) [inline apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.ts:478]: **Abstentions go invisible when `allow_abstain` is false, but still suppress the zero-state**

This returns false whenever `abstainedVoters > 0`, with no reference to the flag — but every display of those abstentions is gated on `voteData.allow_abstain`. For a vote reporting `num_abstained > 0` with the flag false or absent (an organizer editing it off mid-vote), the generic voter final view shows an empty results list with neither the abstain row nor "No votes were submitted", and the card's inflated `totalResponses` has nothing explaining it.

Should the display gate follow the same rule as this short-circuit?

```html
@if (voteData.allow_abstain || participationStats().abstainedVoters > 0) {
```

File: packages/shared/src/utils/vote.utils.ts (hunk 5 above)
Comment (audigregorie) [reply packages/shared/src/utils/vote.utils.ts:148]: Confirmed against upstream: lfx-v2-voting-service passes ITX's percentage through untouched, and its docs/api-contracts.md results example (num_votes_cast: 18, num_abstained: 2, choices 14 → 77.8% and 10 → 55.6% on a multiple-choice question) shows the percentage base is num_votes_cast — the same denominator abstainedRate uses, so API-path numbers agree. Kept abstainedRate as-is (the choice-only base would diverge for multiple-choice: 2/26 vs choices on /18), and in f3177e9 hardened the one divergent path: the computePercentage fallback now also divides by num_votes_cast.

File: apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.ts (hunk 7 above)
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.ts:478]: Fixed in f3177e9 — adopted the suggested rule via a showAbstain computed (allow_abstain || abstainedVoters > 0) that now drives all seven abstain display gates, so a flag-off vote with recorded abstentions shows the row instead of an empty list. Skeletons stay on allow_abstain alone since abstainedVoters is always 0 mid-load.

File: apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html (hunk 6 above)
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html]: Fixed in f3177e9 — all five creator abstain elements (live + final abstain rows and the three Abstained participation stats) now gate on !voteResultsError(), matching the voter-side treatment from 4dce2cf. The broader creator error banner for the pre-existing zeroed stats is left as follow-up scope.

File: apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html [hunk 8]
```diff
@@ -306,6 +306,23 @@ <h4 class="text-sm font-medium text-slate-700">
                   </div>
                 </div>
               }
+              @if (showAbstain() && !voteResultsError()) {
+                <!-- Vote-level abstention row — muted and never winner-styled; choice percentages above exclude abstentions. -->
```
[the remark this answers is not in the document]
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html]: Fixed in 4c3866715 — comment now states choice percentages use all responses cast (including abstentions) as their denominator, matching computePercentage.

File: apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html [hunk 9]
```diff
@@ -534,6 +586,23 @@ <h3 class="text-base font-semibold text-slate-900">Q{{ idx + 1 }}. {{ question.q
                 </div>
               </div>
             }
+
+            @if (showAbstain() && !voteResultsError()) {
+              <!-- Vote-level abstention row (not per-question) — muted and never winner-styled.
+                   abstainedRate is the share of all responses cast; choice percentages above exclude abstentions. -->
```
[the remark this answers is not in the document]
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-results-drawer/vote-results-drawer.component.html]: Fixed in 4c3866715 — both occurrences (lines 592 and 769) now say choice percentages use the same all-responses-cast denominator; only choice counts exclude abstentions.

**Review by ahmedomosanya — APPROVED** [review]
Re-reviewed after the follow-up commits — all three blockers from the earlier review are resolved:

- **Percentage denominator:** `computePercentage` fallback now divides by `num_votes_cast`, matching upstream's documented base for choice percentages and `abstainedRate`.
- **Error gates:** Creator abstain UI (stat + rows) now gates on `!voteResultsError()`, consistent with the voter-side fix.
- **`showAbstain`:** Abstentions render when results report them even if `allow_abstain` was switched off mid-vote, aligned with the `hasZeroVotes` short-circuit.

Form plumbing (`allow_abstain` through all four builders + mapper default), shared `computeVoteParticipationStats`, and the new unit tests all look good. LGTM.

**Review by luismoriguerra — APPROVED** [review]

**Review by MRashad26 — APPROVED** [review]
